robot-server/robot_server/runs/router/commands_router.py

```diff
@@ -93,10 +101,30 @@ async def create_run_command(
             status.HTTP_400_BAD_REQUEST
         )
 
-    command = engine_store.engine.add_command(request_body.data)
+    added_command = engine_store.engine.add_command(request_body.data)
+
+    command_to_return: pe_commands.Command
+
+    if waitUntilComplete:
+        try:
+            await wait_for(
+                engine_store.engine.wait_for_command(command_id=added_command.id),
+                timeout=timeout,
+            )
+            completed_command = engine_store.get_state(run.id).commands.get(
+                command_id=added_command.id
+            )
+            command_to_return = completed_command
+        except AsyncioTimeoutError:
+            timed_out_command = engine_store.get_state(run.id).commands.get(
+                command_id=added_command.id
+            )
+            command_to_return = timed_out_command
+    else:
+        command_to_return = added_command
```

Comment:
See comment above about having `wait_for_command` return the command and pushing timeout down. This then becomes

```suggestion
    engine = engine_store.engine
    command: pe_commands.Command = engine.add_command(request_body.data)

    if waitUntilComplete:
        try:
            command = await engine.wait_for_command(command.id, timeout=timeout)

        except AsyncioTimeoutError:
            command = engine.state_view.commands.get(command.id)

```

Reply:
There's also an argument here that `engine.wait_for_command` shouldn't raise at all, and instead should return the command regardless of if the timeout occurs. Nothing in our code cares that a timeout occurs, so burying it would make this code even simpler and easier to test:

```python
engine = engine_store.engine
command: pe_commands.Command = engine.add_command(request_body.data)
    
if waitUntilComplete:
    command = await engine.wait_for_command(command.id, timeout=timeout)
```

Reply:
Agreed with a lot of this. See discussion in other threads.